### ISSUE TITLE
Add rdf:langString datatype to make "graphs_and_datasets" example work

### ISSUE THREAD

**Issue by liss-h** [TortugaAttack/rdf4cpp#61]
Currently rdf:langString does not exist as a LiteralDatatype, but it is used in the mentioned example as though it is one.
This provokes a bug kind of related to #55 when #56 is merged as during comparisons the type is required to exist in the registry (which it does not) to fetch the corresponding function to perform the actual comparison.

To Fix:

- [x] "Just" define it as a LiteralDatatype
- [x]  Extend #56 to fallback to lexical form comparisons in the extended (order by) comparison operators (which are used for Node comparisons)

**Comment by bigerl** [comment]
I think, we need both: 
- please add `rdf:langString` as `datatypes/rdf/langString.hpp`
- For `order by`, add a fallback to string comparison for types that are not comparable. Please double check that this is not against the SPARQL specs.

**Comment by liss-h** [comment]
Well, modeling a LiteralDatatype for `rdf:langString` is turning out to be quite painful.

Options:
1. cpp_type = `std::string` storing just the data without the language tag
2. cpp_type = something equivalent to `struct { std::string data; std::string language_tag; }`

**TLDR with my opinion is on the bottom**

Problems with 1:
- I think it's kinda weird/unexpected for the `Literal::value()` of a language tagged string to have no way to actually get the language
- `Literal::make<LangString>("hello")` is not possible since there is no way to infer the language tag from that; which is inconsistent
- `Literal::make<LangString>(R"("hello"@en)")` is incorrect since this does not map to the actual representation value

Problems with 2.:
- `LangString::from_string` has to either
  - work differently from any other datatype since `from_string` now has to take `"hello"@en` like data which is not the lexical form of the value
  - take the actual lexical form (`hello`) but then they can't populate/have to throw away the language_tag field, which would make it useless
- `LangString::to_string` has to either
  - throw away the language tag
  - return data like `"hello"@en` which is not the lexical form 
  
  Whatever it is is requires special cases in _every_ single constructor and function operating on the data in `Literal`. And the type behaves completely differently from any other datatype.

The fundamental problem here is that the LiteralDatatypes, Literal and NodeStorage are modeled in a way that makes them ever so slightly incompatible which normally doesn't show up but here it just breaks.

My preferred solution (which is option 2 and added workarounds in literal; but still slightly wrong):
- `Literal::make<LangString>(LangStringRepr{"hello", "en"})` ok
- `Literal{"hello", "en"}` = `Literal::make<LangString>(LangStringRepr{"hello", "en"})`
- `Literal{R"("hello"@en)", rdf_lang_string_iri}` = `Literal{"hello", "en"}` // this constructor is the only thing that is weird
- `Literal{"hello", "en"}.lexical_form()` -> `hello`
- `Literal{"hello", "en"}.value()` -> `LangStringRepr{"hello", "en"}`
- `LangString::from_string(R"("hello"@en)")` -> `LangStringRepr{"hello", "en"}`
- `LangStringRepr{"hello", "en"}.to_string()` -> `"hello"@en`

**Comment by bigerl** [comment]
I prefer option 2. Please double check how jena and rdflib handle it. If you have suggestions how to change the APIs of the node Storage or Literal, bring them on. :)

**Comment by liss-h** [comment]
> I prefer option 2. Please double check how jena and rdflib handle it. If you have suggestions how to change the APIs of the node Storage or Literal, bring them on. :)

**rdflib:**
behaves more like option 1; but allows to enter invalid state without warning
```python
# plain literal
x = Literal("hello", lang="en")
assert type(x.value) == str
assert x.value == "hello"
assert x.language == "en"
assert x.datatype is None

# typed literal
y = Literal("hello", datatype=RDF.langString)
assert str(y) == "hello"
assert y.value is None # no idea why the value is None
assert y.language is None
assert y.datatype == RDF.langString

# both
try:
    Literal("hello", lang="en", datatype=RDF.langString)
except TypeError as e:
    # hint to why it is broken
    assert str(e) == "A Literal can only have one of lang or datatype, per http://www.w3.org/TR/rdf-concepts/#section-Graph-Literal"
```

**jena:**
simmilar
```java
var m = ModelFactory.createDefaultModel();

// plain literal
var lit = m.createLiteral("hello", "en");
assert(lit.getString().equals("hello"));
assert(lit.getLanguage().equals("en"));
assert(lit.getDatatype() == RDFLangString.rdfLangString);

// typed literal; no way to specify lang tag
var lit2 = m.createTypedLiteral("hello", RDFLangString.rdfLangString);
assert(lit2.getString().equals("hello"));
assert(lit2.getLanguage().equals(""));
assert(lit2.getDatatype().equals(RDFLangString.rdfLangString));
```
and RDFLangString is modeled as String
```java
public class RDFLangString extends BaseDatatype implements RDFDatatype {
  ...
  public Object parse(String lexicalForm) {
    return lexicalForm;
  }
}
```

So basically both rdflib and jena behave, at least to me, in an unexpected way. Rdflib at least gave a link to the spec section: a Literal is either a "Plain Literal" with an optional lang tag or a typed literal but never both at the same time. So this isn't really a problem in our modeling but the spec is just weird.. which begs the question why rdf:langString is even specified as a datatype if it's not really a datatype.
Up until now I was thinking that `"hello"@en` was just a weird shorthand for something like `"hello"@en^^rdf:langString` as it turns out it doesn't work that way and this is just an error.

So I guess it's option 1 then. Or alternatively: don't implement it as a datatype at all because it would just create confusing behaviour.

**Comment by bigerl** [comment]
Thank you for working out the details.
I am not convinced by their approach and would still vote for option 2. 
<del>I would still keep it as a datatype because it is defined as such by RDF.</del>

At least a dummy datatype would be good. Implementation details can also be handled e.g. in Literal.

**Comment by liss-h** [comment]
#63

**Comment by liss-h** [comment]
merged
